existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,292 @@
+In order to do it, you pass a logging function as an argument.
+
+```haskell
+func :: _ -> IO ()
+func log = do
+  username <- getUsername
+  log username
+
+  userCount <- getUserCount
+  log userCount
+
+getUsername :: IO String
+getUsername = undefined
+
+getUserCount :: IO Int
+getUserCount = undefined
+```
+
+What type should the log function have?
+Since it somehow logs both `String` and `Int`:
+
+```haskell
+log :: Show a => a -> IO ()
+```
+However, if we add this signature to `func` and try to call it with different arguments:
+
+```haskell
+func :: Show a => (a -> IO ()) -> IO ()
+func log = ...
+
+main :: IO ()
+main = do
+  -- log to the console
+  func print
+
+  -- log to a logfile
+  func (appendFile "logfile.log" . show)
+```
+We will see some nasty compiler errors:
+```haskell
+  log username
+      ^^^^^^^^
+Couldn't match expected type `a` with actual type `String`...
+```
+
+We're getting this error because:
+1. `func`'s type variable `a` is universally quantified, therefore it can only be instantiated at the use-site.
+2. We're trying to instantiate it at `func`'s definition-site (`log username` is trying to instantiate the type variable `a` with `String` and `log userCount` is trying to instantiate the type variable `b` with `Int`).
+Because we want `a` to be instantiated at `func`'s definition-site (rather than its use-site), we need `a` to be existentially quantified instead.
+
+```haskell
+{-# LANGUAGE RankNTypes #-}
+
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+Now the compiler knows, that whatever function we pass to `func`, type variable `a` should be instantiated here, not outside.
+With this change we can instantiate `a` when we call log.
+
+```haskell
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+func log = do
+  username <- getUsername
+  log @String username
+
+  userCount <- getUserCount
+  log @Int userCount
+```
+
+Note, that this type applications is unnecessary.
+Compiler is smart enough to choose appropriate types.
+
+#### Heterogeneous lists
+
+Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
+Let's take a look how this is done in datatypes.
+You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of the constructor name. Existential types in datatypes allow you to have field in the datatype, without necessity to specify its type.
+
+```haskell
+data Elem = forall a. Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+```
+
+This is a heterogeneous list that contains values of different types.
+You can't do that with an universally quantified list because every element in that list would need to have the same type.
+But, there are not many things you can do with elements of this list.
+Since the type of elements is unknown at the use-site, you can't address it, nor can you pass it to functions that expect values of concrete type.
```

Comment:
> Since the type of elements is unknown at the use-site, you can't address it, nor can you pass it to functions that expect values of concrete type.

Maybe here or below, we can show the ability of `Typeable` typeclass to bring back concrete type into scope, enabling these values to be passed to functions that expect concrete types.

Reply:
done